src/test/java/com/android/volley/AsyncRequestQueueTest.java

```diff
@@ -0,0 +1,159 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.volley;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+import static org.mockito.MockitoAnnotations.initMocks;
+
+import com.android.volley.mock.ShadowSystemClock;
+import com.android.volley.toolbox.NoCache;
+import com.android.volley.toolbox.StringRequest;
+import com.android.volley.utils.ImmediateResponseDelivery;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+/** Unit tests for RequestQueue, with all dependencies mocked out */
+@RunWith(RobolectricTestRunner.class)
+@Config(shadows = {ShadowSystemClock.class})
+public class AsyncRequestQueueTest {
+
+    private ResponseDelivery mDelivery;
+    @Mock private AsyncNetwork mMockNetwork;
+
+    @Before
+    public void setUp() throws Exception {
+        mDelivery = new ImmediateResponseDelivery();
+        initMocks(this);
+    }
+
+    @Test
+    public void cancelAll_onlyCorrectTag() throws Exception {
+        AsyncRequestQueue queue =
+                new AsyncRequestQueue.Builder(mMockNetwork)
+                        .setCache(new NoCache())
+                        .setResponseDelivery(mDelivery)
+                        .build();
+        queue.start();
+        Object tagA = new Object();
+        Object tagB = new Object();
+        StringRequest req1 = mock(StringRequest.class);
+        when(req1.getTag()).thenReturn(tagA);
+        StringRequest req2 = mock(StringRequest.class);
+        when(req2.getTag()).thenReturn(tagB);
+        StringRequest req3 = mock(StringRequest.class);
+        when(req3.getTag()).thenReturn(tagA);
+        StringRequest req4 = mock(StringRequest.class);
+        when(req4.getTag()).thenReturn(tagA);
+
+        queue.add(req1); // A
+        queue.add(req2); // B
+        queue.add(req3); // A
+        queue.cancelAll(tagA);
+        queue.add(req4); // A
+
+        verify(req1).cancel(); // A cancelled
+        verify(req3).cancel(); // A cancelled
+        verify(req2, never()).cancel(); // B not cancelled
+        verify(req4, never()).cancel(); // A added after cancel not cancelled
+        queue.stop();
+    }
+
+    @Test
+    public void add_notifiesListener() throws Exception {
+        RequestQueue.RequestEventListener listener = mock(RequestQueue.RequestEventListener.class);
+        AsyncRequestQueue queue =
+                new AsyncRequestQueue.Builder(mMockNetwork)
+                        .setCache(new NoCache())
```

Comment:
should we define an Async equivalent of NoCache and use it here to exercise the Async flow?

Reply:
Yeah, I think that would be a good idea. Would that just look like a method that calls the callback with null?